src/browser/util/url.js

```diff
@@ -10,87 +10,64 @@ define(function() {
 
   'use strict';
 
-  var rscheme = /^(?:[a-z\u00a1-\uffff0-9-+]+)(?::|:\/\/)/i;
+  const rscheme = /^(?:[a-z\u00a1-\uffff0-9-+]+)(?::|:\/\/)/i;
 
-  var UrlHelper = {
-
-    // Placeholder anchor tag to format URLs.
-    a: null,
-
-    getUrlFromInput: function urlHelper_getUrlFromInput(input) {
-      this.a = this.a || document.createElement('a');
-      this.a.href = input;
-      return this.a.href;
-    },
+  const UrlHelper = {
 
     hasScheme: function(input) {
       return !!(rscheme.exec(input) || [])[0];
     },
 
-    isURL: function urlHelper_isURL(input) {
-      return !UrlHelper.isNotURL(input);
-    },
-
     getOrigin: function urlHelper_getOrigin(url) {
-      this.a = this.a || document.createElement('a');
-      this.a.href = url;
-      return this.a.origin;
+      return new URL(url).origin;
     },
 
     getBaseURI: function urlHelper_getBaseURI() {
-      // http://example.com/foo/index.html -> http://example.com/foo/
-      // http://example.com/foo/ -> http://example.com/foo/
-      return location.href.replace(/\/[^/]*$/, '/');
+      return new URL('./', location);
     },
 
     getHostname: function urlHelper_getHostname(url) {
-      this.a = this.a || document.createElement('a');
-      this.a.href = url;
-      return this.a.hostname;
+      return new URL(url).hostname;
     },
+
     getDomainName(url) {
       return this.getHostname(url).replace(/^www\./, '');
     },
 
     getProtocol: function urlHelper_getProtocol(url) {
-      this.a = this.a || document.createElement('a');
-      this.a.href = url;
-      return this.a.protocol;
+      return new URL(url).protocol;
     },
 
+
     isNotURL: function urlHelper_isNotURL(input) {
-      var schemeReg = /^\w+\:\/\//;
+      let str = input.trim();
 
-      // in bug 904731, we use <input type='url' value=''> to
-      // validate url. However, there're still some cases
-      // need extra validation. We'll remove it til bug fixed
-      // for native form validation.
-      //
       // for cases, ?abc and 'a? b' which should searching query
-      var case1Reg = /^(\?)|(\?.+\s)/;
+      const case1Reg = /^(\?)|(\?.+\s)/;
       // for cases, pure string
-      var case2Reg = /[\?\.\s\:]/;
+      const case2Reg = /[\?\.\s\:]/;
       // for cases, data:uri
-      var case3Reg = /^(data\:)/;
+      const case3Reg = /^(data\:)/;
       // for cases, only scheme but no domain provided
-      var case4Reg = /^\w+\:\/*$/;
-      var str = input.trim();
+      const case4Reg = /^\w+\:\/*$/;
+      if (str == 'localhost') {
+        return false;
+      }
       if (case1Reg.test(str) || !case2Reg.test(str) || case4Reg.test(str)) {
         return true;
       }
       if (case3Reg.test(str)) {
         return false;
       }
```

Comment:
do we really need case3Reg at all? data: urls should be recognized properly by new URL()